storage/watchable_store_test.go

```diff
@@ -59,6 +58,151 @@ func TestNewWatcherCancel(t *testing.T) {
 	}
 }
 
+// TestCancelUnsynced tests if running CancelFunc removes watchings from
+// synced and unsynced.
+func TestCancelUnsynced(t *testing.T) {
+	// manually create watchableStore instead of newWatchableStore
+	// because newWatchableStore automatically calls syncWatchers
+	// method to sync watchers in unsynced map. We want to keep watchers
+	// in unsynced to test if syncWatchers works as expected.
+	s := &watchableStore{
+		store:    newStore(tmpPath),
+		unsynced: make(map[*watching]struct{}),
+
+		// to make the test not crash from assigning to nil map.
+		// 'synced' doesn't get populated in this test.
+		synced: make(map[string]map[*watching]struct{}),
+	}
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	// Put a key so that we can spawn watchers on that key.
+	// (testKey in this test). This increases the rev to 1,
+	// and later we can we set the watcher's startRev to 1,
+	// and force watchers to be in unsynced.
+	testKey := []byte("foo")
+	testValue := []byte("bar")
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	// create watcherSize of CancelFunc of
+	// synced and unsynced
+	cancels := make([]CancelFunc, 2*watcherSize)
+	for i := 0; i < watcherSize; i++ {
+		// use 1 to keep watchers in unsynced
+		_, cancel := w.Watch(testKey, true, 1)
+		cancels[i] = cancel
+	}
+	for i := watcherSize; i < 2*watcherSize; i++ {
+		// use 0 to keep watchers in synced
```

Comment:
what i really hope is that we ONLY test unsynced case in this test. Since This test is called `TestCancelUnsynced`. We do not test anything about sync.